### ISSUE TITLE
u_iscntrl like helper?

### ISSUE THREAD

**Issue by michaelsbradleyjr** [dankamongmen/notcurses#2524]
Would it make sense for notcurses to provide a helper like [`u_iscntrl`](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/uchar_8h.html#ab1648dc8c38032549e7675dd697ad24b)? (for impl see [icu4c/source/common/uchar.cpp#L182-L188](https://github.com/unicode-org/icu/blob/2dce62892b6b00df02195a35ca09ebd8e082e80f/icu4c/source/common/uchar.cpp#L182-L188))?

The idea is that it could be used as a basis for determining whether to substitute a codepoint (as utf8) with e.g. [`0x10fFFEE`](https://codepoints.net/U+10FFEE?lang=en) (as utf8), aka ["tofu"](https://en.wikipedia.org/wiki/Unicode_input#Availability), before passing it to `ncplane_putstr`.

Maybe notcurses already provides an equivalent facility and I haven't found it yet? Or what I have in mind doesn't make sense in the context of notcurses (for reasons I don't understand yet)?

**Comment by dankamongmen** [comment]
so basically something that could explicitly test whether we can put an EGC through `write()` (this wouldn't guarantee that there's actually a font element for it).

let me ask: what would you do if you found part of your text couldn't be printed? i guess it depends on the text (look around in the source tree for places where we test whether UTF8 is available, and otherwise special-case for ASCII, a tres lame practice).

this is trivial to wrap libunistring for, but i'm not sure that correlates perfectly with what generates an error from `ncplane_put()` (it probably does not, since libunistring uses its own tables AFAIK).

just making sure i understand your req.

**Comment by michaelsbradleyjr** [comment]
I'm not sure if I understand my own request. 😄 

Here's the context of my question:

This is my adaptation of Notcurses `cli1` in nim-notcurses:
https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/examples/poc/cli1.nim

Note that keys aren't actually reported yet, but will be soon, just need to wire up some constants to an enum in the Nim API, but that's another matter.

Per L38, utf8 stuff that is input to the terminal will get rendered as `utf8  : [stuff]`, character by character.

So, if I copy/paste e.g. [`U+0080`](https://codepoints.net/U+0080?lang=en) into Kitty or iTerm2 or whatever, I'm getting a correct report re: `ncinput`'s fields: `id` is `128`, the utf8 array is `['\xC2', '\x80', '\x00', '\x00', '\x00']`, etc.

What I found is that if I try to do e.g. `stdn.putStr("utf8 : " & utf8.get & "\n")`, where `utf8.get` is returning the utf8-encoded string for `U+0080`, then `ncplane_putstr` returns a negative value, i.e. a failure if, I understand correctly.

However, if I do `stdn.putStr("utf8 : ")` (and render), and then `stdn.putStr(utf8.get)` (and render), `ncplane_putstr` returns a positive value for both puts; *but* when I putstr newline and then render, there's no additional visible output on that line and if I use the mouse to select text on that line, there's nothing it picks up after `utf8 : `.

The same happens for other codepoints like those for escape, tab, and so on. The behavior is the same in multiple terminals.

So I started researching, eventually hit on `u_iscntrl` and thus my question here.

The way I'm going about this may be all wrong, so mainly I'm trying to form the right ideas about Unicode and UTF-8 in the context of Notcurses.

**Comment by michaelsbradleyjr** [comment]
Here's a screenshot of Terminal.app re: what I described above. After starting my `cli1`, I got `U+0080` into my clipboard from codepoints.net and pasted it into the terminal, then I pressed `a`, then I pressed the tab key.
![image](https://user-images.githubusercontent.com/194260/148157710-722303d8-4888-4ec5-b0d7-98bde51caa34.png)

**Comment by dankamongmen** [comment]
> returns a positive value for both puts; but when I putstr newline and then render, there's no additional visible output on that line after utf8 : ; and if I use the mouse to select text on that line, there's nothing it picks up after utf8 : .

what would you expect to happen here?

> What I found is that if I try to do e.g. stdn.putStr("utf8 : " & utf8.get & "\n"), where utf8.get is returning the utf8-encoded string for U+0080, then ncplane_putstr returns a negative value, i.e. a failure if, I understand correctly

is the plane in scrolling mode? you ought only be able to emit newlines to scrolling planes. that you seem able to emit them one way but not another way is troubling in any mode, though.

**Comment by michaelsbradleyjr** [comment]
> what would you expect to happen here?

I'm not sure, I'm only now building intuition about these things.

I was surprised that the put for only `U+0080` returned a positive value while the put for the combined string resulted in a negative return. Is that expected?

> is the plane in scrolling mode?

Yes, the import on [L1](https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/examples/poc/cli1.nim#L1) is `import notcurses/cli` whereby `Nc.init` on [L5](https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/examples/poc/cli1.nim#L5) initializes Notcurses in CLI mode:
https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/notcurses/api/private/cli/init.nim#L4-L6
https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/notcurses/api/private/cli/init.nim#L22

**Comment by dankamongmen** [comment]
oh, i was literally just like, "what do you think ought happen here"? so when a plane is scrolling, emitting a newline is allowed, and ought:

1) move the cursor down, if we are not on the last line, or
2) scroll, if we are, and
2) move the cursor to the start of the line

but it never emits an actual character.

UNLESS

you are emitting it to the standard plane, and the standard plane is scrolling. in this case it can translate to an actual escape sequence, but never one you can see or copy-and-paste.

**Comment by dankamongmen** [comment]
> oh, i was literally just like, "what do you think ought happen here"?

i.e. i am steeped in notcurses approach, and i am interested in the fresh expectations someone else has *before* they develop fluency

**Comment by michaelsbradleyjr** [comment]
> i.e. i am steeped in notcurses approach, and i am interested in the fresh expectations someone else has before they develop fluency

I took it that way, for sure, and in this case I'm so fresh to it that I don't have many expectations at all. Currently for me, the experience is more like:

![image](https://user-images.githubusercontent.com/194260/148163991-4a68c4fd-7e55-4920-85e6-ea5f8f0f033d.png)

What you explained in 1-3 above is helpful. I'm trying to understand the behavior of `ncplane_putstr`. I'm going to use some pseudocode (of sorts).

Why does:
```
ncplane_putstr("utf8 : " + u0080 + "\n")
```
return a negative number? While in this case:
```
ncplane_putstr("utf8 : ")
render()
ncplane_putstr(u0080)
render()
ncplane_putstr("\n")
```
all three calls of `ncplane_putstr` return positive numbers?

Put another way, why is put okay with `U+0080` by itself but not when it's within a longer string?

`notcurses.h` says for `ncplane_putstr_yx`, which is called by `ncplane_putstr`:
```
// Write a series of EGCs to the current location, using the current style.
// They will be interpreted as a series of columns (according to the definition
// of ncplane_putc()). Advances the cursor by some positive number of columns
// (though not beyond the end of the plane); this number is returned on success.
// On error, a non-positive number is returned, indicating the number of columns
// which were written before the error.
```

The answer may already be in your 1-3 above and I just didn't grok it.

One of the things I'm attempting with my Nim API is to have strong consistency around use of `Result[Value, Error]` and `Option[T]` for return types, i.e. where it makes sense. `Result[V, E]` is basically the Error monad and `Option[T]` the Maybe monad, if you're fluent in monad[^1].

The point being that I want and need to develop a strong intuition about errors, successes, and values in Notcurses so I can make good decisions re: if/when they can fit in those type patterns.

[^1]: Most Nim programmers probably aren't, but the patterns that arise from those types are still pretty user friendly and helpful without users having to know anything about monads.

**Comment by michaelsbradleyjr** [comment]
After some more experiments, my question is, per:
```
On error, a non-positive number is returned...
```
I should consider `0` to be non-positive?

If the answer is *yes* then I think I understand what/how I need to change the Nim API and my `cli1` to that I can deal with this case consistently.

**Comment by dankamongmen** [comment]
this clearly needs some looking into; i'll get into it this weekend at the latest.

**Comment by michaelsbradleyjr** [comment]
Fwiw, I changed my API around `ncplane_putstr` so that it checks `<= 0` instead of `< 0`:

https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/notcurses/api/private.nim#L113-L116

And with that change, I was able to modify my `cli1` logic to work with that:

https://github.com/michaelsbradleyjr/nim-notcurses/blob/version_3_revamp/examples/poc/cli1.nim#L54-L56

It's straightforward in this case because it's a matter of putting just one character (plus newline). So if the put failed then I know to put the replacement character plus newline.

There's still the more general case of escaping one/more non-put-able codepoints (otherwise valid utf8) that might be anywhere in a longer string. However, maybe that's a thing that's not really a concern in practice.

I'll leave the issue open since you feel it needs looking into. For my part, I feel I misunderstood "non-positive" to mean "negative" strictly speaking. Since `0` cols get written when I do `putstring(u0080 + "\n")` and it fails, it makes sense that the error code is `0` since that code represents the number of columns written before the error.
